# Telco Customer Churn Analysis

## Overview
This analysis explores customer churn behavior for a telecom company using the Telco Customer Churn dataset.
The goal is to identify why customers are leaving and provide actionable recommendations to reduce churn.
## Dataset
- Total customers: 7,032
- Features: 21 columns covering customer demographics, services, contract details and billing
- Target column: `Churn` — whether the customer left the company or not


In [9]:
from sqlalchemy import create_engine
import pandas as pd

engine = create_engine("mysql+mysqlconnector://root:aqsam1022@localhost/churn")

def query(query):
    return pd.read_sql(query, engine)

In [10]:
# 1. overall churn rate
query("""
    SELECT Churn,
           COUNT(*) as total_customers,
           ROUND(COUNT(*) * 100.0 / SUM(COUNT(*)) OVER(), 1) as percentage
    FROM customer
    GROUP BY Churn
""")

,Churn,total_customers,percentage
0,No,5163,73.4
1,Yes,1869,26.6


## Finding 1 - Overall Churn Rate
- **26.6%** of customers churned which is roughly 1 in 4 customers is leaving
- 1,869 customers lost out of 7,032 total

In [12]:
# 2. contract type distribution among churned customers only
query("""
    SELECT Contract,
           COUNT(*) as churned_count,
           ROUND(COUNT(*) * 100.0 / SUM(COUNT(*)) OVER(), 1) as percentage
    FROM customer
    WHERE Churn = 'Yes'
    GROUP BY Contract
    ORDER BY churned_count DESC
""")

,Contract,churned_count,percentage
0,Month-to-month,1655,88.6
1,One year,166,8.9
2,Two year,48,2.6


## Finding 2 - Contract Type is the Biggest Driver
- **88.6%** of all churned customers were on **month-to-month** contracts
- One year and two year contracts together account for only 11.4% of churn

In [29]:
# 3. avg monthly charge by contract type
query("""
    SELECT Contract,
           Count(case when Churn = 'Yes' then 1 end) as churned,
           Count(case when Churn = 'No' then 1 end) as Stayed,
           COUNT(*) as total_customers,
           Count(case when Churn = 'Yes' then 1 end)* 100 / count(*) as churned_percentage,
           ROUND(AVG(MonthlyCharges), 2) as avg_monthly_charge
           
    FROM customer
    GROUP BY Contract
    ORDER BY avg_monthly_charge DESC
""")

,Contract,churned,Stayed,total_customers,churned_percentage,avg_monthly_charge
0,Month-to-month,1655,2220,3875,42.7097,66.40
1,One year,166,1306,1472,11.2772,65.08
2,Two year,48,1637,1685,2.8487,60.87


Key insight: The average monthly charge difference between contract types is small ($66 vs $65 vs $60) but the churn difference is massive. This tells us the problem is **lack of commitment**, not price.


In [15]:
# 4. churn rate by internet service and contract type
query("""
    SELECT Contract,
           InternetService,
           COUNT(*) as total,
           SUM(CASE WHEN Churn = 'Yes' THEN 1 ELSE 0 END) as churned,
           ROUND(SUM(CASE WHEN Churn = 'Yes' THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) as churn_rate_pct,
           ROUND(AVG(MonthlyCharges), 2) as avg_monthly_charge
    FROM customer
    GROUP BY Contract, InternetService
    ORDER BY churn_rate_pct DESC
""")

,Contract,InternetService,total,churned,churn_rate_pct,avg_monthly_charge
0,Month-to-month,Fiber optic,2128,1162.0,54.6,87.02
1,Month-to-month,DSL,1223,394.0,32.2,50.22
2,One year,Fiber optic,539,104.0,19.3,98.78
3,Month-to-month,No,524,99.0,18.9,20.41
4,One year,DSL,570,53.0,9.3,61.40
5,Two year,Fiber optic,429,31.0,7.2,104.57
6,One year,No,363,9.0,2.5,20.82
7,Two year,DSL,623,12.0,1.9,70.51
8,Two year,No,633,5.0,0.8,21.77


Finding 3 - Internet Service Drives Churn Within Month-to-Month
- Fiber optic customers on month-to-month churn at **54.6%**
- DSL customers on month-to-month churn at **32.2%**
- But two year fiber optic customers churn at only **7.2%** even though they pay more ($104/month)

In [30]:
query("""
    SELECT 
        CASE 
            WHEN tenure BETWEEN 1 AND 2 THEN '01-02 months'
            WHEN tenure BETWEEN 3 AND 5 THEN '03-05 months'
            WHEN tenure BETWEEN 6 AND 10 THEN '06-10 months'
            WHEN tenure BETWEEN 11 AND 15 THEN '11-15 months'
            WHEN tenure BETWEEN 16 AND 20 THEN '16-20 months'
            WHEN tenure BETWEEN 21 AND 30 THEN '21-30 months'
            WHEN tenure BETWEEN 31 AND 40 THEN '31-40 months'
            WHEN tenure BETWEEN 41 AND 50 THEN '41-50 months'
            WHEN tenure BETWEEN 51 AND 60 THEN '51-60 months'
            WHEN tenure BETWEEN 61 AND 72 THEN '61-72 months'
        END as tenure_bucket,
        COUNT(*) as churned_count,
        COUNT(CASE WHEN Churn = 'No' THEN 1 END) as stayed_count,
        ROUND(COUNT(CASE WHEN Churn = 'Yes' THEN 1 END) * 100.0 / COUNT(*), 1) as churn_pct
    FROM customer
    GROUP BY tenure_bucket
    ORDER BY tenure_bucket
""")

,tenure_bucket,churned_count,stayed_count,churn_pct
0,01-02 months,851,348,59.1
1,03-05 months,509,268,47.3
2,06-10 months,599,375,37.4
3,11-15 months,500,332,33.6
4,16-20 months,408,293,28.2
5,21-30 months,763,589,22.8
6,31-40 months,645,504,21.9
7,41-50 months,652,537,17.6
8,51-60 months,698,603,13.6
9,61-72 months,1407,1314,6.6


## Finding 4 - Tenure Threshold
- Churn drops consistently as tenure increases
- Customers in months **1-2 churn at 59.1%** — the highest risk window
- By months **61-72 churn drops to just 6.6%**

In [32]:
query("""
    SELECT 
        InternetService,
        COUNT(*) as total_month1_customers,
        SUM(CASE WHEN Churn = 'Yes' THEN 1 ELSE 0 END) as month1_churned,
        SUM(CASE WHEN Churn = 'No' THEN 1 ELSE 0 END) as month1_stayed,
        ROUND(SUM(CASE WHEN Churn = 'Yes' THEN 1 ELSE 0 END) * 100.0 / COUNT(*), 1) as churn_pct,
        ROUND(AVG(CASE WHEN Churn = 'Yes' THEN MonthlyCharges END), 2) as avg_charge_churned,
        SUM(CASE WHEN Churn = 'Yes' AND TechSupport = 'Yes' THEN 1 ELSE 0 END) as tech_support_yes,
        SUM(CASE WHEN Churn = 'Yes' AND TechSupport = 'No' THEN 1 ELSE 0 END) as tech_support_no
    FROM customer
    WHERE tenure = 1
    GROUP BY InternetService
    ORDER BY month1_churned DESC
""")

,InternetService,total_month1_customers,month1_churned,month1_stayed,churn_pct,avg_charge_churned,tech_support_yes,tech_support_no
0,Fiber optic,235,203.0,32.0,86.4,77.76,3.0,200.0
1,DSL,212,119.0,93.0,56.1,43.40,11.0,108.0
2,No,166,58.0,108.0,34.9,20.08,0.0,0.0


## Finding 5 - Month 1 is the Most Critical Period
- **375 customers** left in month 1 alone, more than any other single month
- Of 235 fiber optic customers who joined, **203 left in month 1** , an 86.4% churn rate
- Of those 203 fiber optic month-1 churners, **200 had no tech support**
> Key insight: New fiber optic customers sign up for the most expensive service, have no support when something goes wrong, and leave within the first month before they even settle in.

In [21]:
# 7. revenue impact
query("""
    SELECT 
        COUNT(*) as total_churned,
        ROUND(SUM(MonthlyCharges), 0) as monthly_revenue_lost,
        ROUND(SUM(MonthlyCharges) * 12, 0) as yearly_revenue_lost,
        ROUND(AVG(MonthlyCharges), 2) as avg_charge_per_churned_customer
    FROM customer
    WHERE Churn = 'Yes'
""")

,total_churned,monthly_revenue_lost,yearly_revenue_lost,avg_charge_per_churned_customer
0,1869,139131.0,1669570.0,74.44


In [22]:
# 8. fiber optic month 1 no tech support — potential saving
query("""
    SELECT 
        COUNT(*) as customers_at_risk,
        ROUND(SUM(MonthlyCharges), 0) as monthly_revenue_at_risk,
        ROUND(SUM(MonthlyCharges) * 12, 0) as yearly_revenue_at_risk
    FROM customer
    WHERE tenure = 1 
      AND Churn = 'Yes'
      AND InternetService = 'Fiber optic'
      AND TechSupport = 'No'
""")

,customers_at_risk,monthly_revenue_at_risk,yearly_revenue_at_risk
0,200,15533.0,186396.0


## Recommendations

**1. Bundle free tech support for new fiber optic customers (first 3 months)**
200 out of 203 fiber optic customers who left in month 1 had no tech support.
A free trial costs less than losing $186,396 per year from this group alone.

**2. Incentivize month-to-month customers to switch to annual contracts**
Month-to-month customers churn at 42.7% vs 2.8% for two year contracts.
Even a small discount to move customers to annual plans would dramatically reduce churn.

**3. Create an early warning system for months 1-5**
Flag all new customers on fiber optic with no add-ons as high risk.
Proactively reach out with support offers before they decide to leave.


## Conclusion
The churn problem is not about price, two year fiber optic customers pay the most and churn the least.
The real issue is new customers on flexible contracts getting a poor first experience with no support.
Fixing the first month experience for fiber optic customers is the single highest impact intervention available.